src/mongocxx/test/database.cpp

```diff
@@ -492,4 +492,60 @@ TEST_CASE("Database integration tests", "[database]") {
     }
 }
 
+// As C++11 lacks generic lambdas, and "ordinary" templates can't appear at block scope,
+// we'll have to define our helper for the serviceId tests here. This implementation would
+// be more straightforward in newer versions of C++ (C++14 and on allow generic lambdas), 
+// see CXX-2350 (migration to more recent C++ standards):
+template <typename EventT>
+struct check_service_id
+{
+    void operator()(const EventT& event) {
+
+    // We MUST NOT report a service_id() outside of load-balancing mode, but in
+    // load-balancing mode, a value is required:
+    INFO("checking for service_id()")
+    CAPTURE(event.command_name(), test_util::is_load_balanced());
+    auto service_id = event.service_id();
+    if (test_util::is_load_balanced())
+        CHECK(service_id);
+    else
+        CHECK_FALSE(service_id);
+    }
+};
+
+
+TEST_CASE("Test serviceId is included in command monitoring events") {
```

Comment:
I suggest splitting this into two `TEST_CASE` blocks.

1. Test mocking each C driver function to return a service ID. Expect each of the C++ callbacks to include `service_id`.
2. Test mocking each C driver function to not return a service ID. Expect each of the C++ callbacks to not include `service_id`.

(2) does not strictly require mocking. All current integration tests in Evergreen run against non load balanced servers (i.e. they will not report a `serviceId` in the handshake). When we do add tests against load balanced servers, we'd have to update (2) to check based on the environment.

Reply:
This is in effect what is happening in my latest change, because of the behavior of GENERATE, which creates virtual SECTIONS. (see: "https://github.com/catchorg/Catch2/blob/devel/docs/generators.md").

Reply:
Ah nice! I did not know about `GENERATE`.